executor/asyncloaddata/util.go

```diff
@@ -178,6 +179,95 @@ func FailJob(
 	return err
 }
 
+// CancelJob cancels a load data job. Only a running/paused job can be canceled.
+func CancelJob(
+	ctx context.Context,
+	conn sqlexec.SQLExecutor,
+	jobID int64,
+	user string,
+) (err error) {
+	ctx = util.WithInternalSourceType(ctx, kv.InternalLoadData)
+	_, err = conn.ExecuteInternal(ctx, "BEGIN PESSIMISTIC;")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			_, err1 := conn.ExecuteInternal(ctx, "ROLLBACK;")
+			terror.Log(err1)
+			return
+		}
+
+		_, err = conn.ExecuteInternal(ctx, "COMMIT;")
+		if err != nil {
+			return
+		}
+	}()
+
+	var (
+		rs   sqlexec.RecordSet
+		rows []chunk.Row
+	)
+	rs, err = conn.ExecuteInternal(ctx,
+		`SELECT expected_status, end_time, error_message FROM mysql.load_data_jobs
+		WHERE job_id = %? AND create_user = %?;`,
+		jobID, user)
+	if err != nil {
+		return err
+	}
+	defer terror.Call(rs.Close)
+	rows, err = sqlexec.DrainRecordSet(ctx, rs, 1)
+	if err != nil {
+		return err
+	}
+
+	if len(rows) < 1 {
+		return exeerrors.ErrLoadDataJobNotFound.GenWithStackByArgs(jobID)
+	}
+	status := rows[0].GetEnum(0).String()
+	if status != "running" && status != "paused" {
+		return exeerrors.ErrLoadDataInvalidOperation.GenWithStackByArgs(fmt.Sprintf("need status running or paused, but got %s", status))
+	}
+	endTimeIsNull := rows[0].IsNull(1)
+	if !endTimeIsNull {
+		hasError := !rows[0].IsNull(2)
+		if hasError {
+			return exeerrors.ErrLoadDataInvalidOperation.GenWithStackByArgs("need status running or paused, but got failed")
```

Comment:
make `failed` a status in job table?

Reply:
currently we only have expected_status. but no actual_status, so these lines is to calculate actual_status from other columns. I plan to add actual_status later when implement PAUSE/RESUME LOAD DATA JOBS, to let PAUSE don't return until actual_status is paused :joy: